File: apple/RemoteDisplay/Views/Display/DisplayViewModel.swift
//
//  DisplayController.swift
//  RemoteDisplay
//
//  Created by Júlio Flores on 22/08/2026.
//

import SwiftUI

@Observable @MainActor
final class DisplayViewModel: Keyboard.Client {
	let id	= UUID()
	nonisolated private let keyboardController: Keyboard.Controller
	private var positions: [Int] = [] // current position for each line

	private(set) var display = Display()

	init(keyboardController: Keyboard.Controller = App.keyboarController) {
		self.keyboardController = keyboardController
		self.keyboardController.subscribe(client: self)
	}

	deinit {
		let clientID = self.id
		let controller = self.keyboardController
		Task { @MainActor in
			controller.unsubscribe(clientID: clientID)
		}
	}

	func setText(_ text: String) {
		self.positions = []
		self.display = Display()
		self.insertText(text)
	}

	func receive(action: Keyboard.Action) {
		switch action {
			case let .text(text):
				self.insertText(text)
			case .backspace:
				self.deleteBackward()
		}
	}

	private var currentCell: Display.Line.Cell {
		get {
			self.display.lines[self.positions.lastIndex].cells[self.positions.last]
		}
		set {
			self.display.lines[self.positions.lastIndex].cells[self.positions.last] = newValue
		}
	}

	private func deleteBackward() {
		guard !positions.isEmpty && positions[0] > 0 else {
			return
		}

		if self.positions.last < Display.Line.Specs.charCount {
			currentCell = .blank
		}
		self.positions.last -= 1
		if self.positions.last < 0 {
			self.positions.removeLast()
			if self.positions.last >= Display.Line.Specs.charCount {
				self.positions.last = Display.Line.Specs.charCount - 1
			}
		}
		if self.positions.lastIndex < Display.Specs.lineCount && self.positions.last < Display.Line.Specs.charCount {
			currentCell = .cursor
		}
	}

	private func insertText(_ text: String) {
		if self.positions.isEmpty {
			self.positions.append(0)
		}
		for c in text {
			if c.isNewline {
				if self.positions.count < Display.Specs.lineCount && self.positions.last > 0 {
					currentCell = .blank
					self.positions.append(0)
				}
			} else if c.isASCII {
				guard self.positions.lastIndex < Display.Specs.lineCount && self.positions.last < Display.Line.Specs.charCount else {
					break
				}
				currentCell = .char(c)
				self.positions.last += 1
				if self.positions.last >= Display.Line.Specs.charCount && self.positions.count < Display.Specs.lineCount {
					self.positions.append(0)
				}
			}
		}
		if self.positions.lastIndex < Display.Specs.lineCount && self.positions.last < Display.Line.Specs.charCount {
			currentCell = .cursor
		}
	}
}

private extension Array {
	var last: Element {
		get {
			self[self.count - 1]
		}
		set {
			self[self.count - 1] = newValue
		}
	}
	var lastIndex: Index {
		self.count - 1
	}
}

File: android/app/src/main/java/dk/cipher/remotedisplay/views/display/DisplayViewModel.kt
package dk.cipher.remotedisplay.views.display

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import dk.cipher.remotedisplay.keyboard.Keyboard
import dk.cipher.remotedisplay.models.Cell
import dk.cipher.remotedisplay.models.Display
import dk.cipher.remotedisplay.models.Line

class DisplayViewModel(val keyboardController: Keyboard.Controller): ViewModel(), Keyboard.Client {
    companion object {
        fun build(keyboardController: Keyboard.Controller) =
            viewModelFactory {
                initializer {
                    DisplayViewModel(keyboardController)
                }
            }
    }

    var display = Display()
    private val positions = mutableListOf<Int>()
    private var currentCell: Cell
        get() {
            return display.lines[positions.lastIndex].cells[positions.last()].value
        }
        set(newValue) {
            display.lines[positions.lastIndex].cells[positions.last()].value = newValue
        }

    init {
        this.keyboardController.subscribe(this)
    }

    override fun onCleared() {
        this.keyboardController.unsubscribe(this)
    }

    override fun receive(action: Keyboard.Action) {
        when (action) {
            is Keyboard.Action.Text -> this.insertText(action.text)
            is Keyboard.Action.Backspace -> this.deleteBackward()
        }
    }
    fun setText(text: CharSequence) {
        positions.clear()
        display = Display()
        insertText(text)
    }

    private fun insertText(text: CharSequence) {
        if (positions.isEmpty()) {
            positions.add(0)
        }
        for (c in text) {
            if (c.isNewLine()) {
                if (positions.size < Display.Specs.lineCount && positions.last() > 0) {
                    currentCell = Cell.Blank
                    positions.add(0)
                }
            } else if (c.isAscii()) {
                if (positions.lastIndex >= Display.Specs.lineCount || positions.last() >= Line.Specs.charCount) {
                    break
                }
                currentCell = Cell.Character(c)
                positions[positions.lastIndex] += 1
                if (positions.last() >= Line.Specs.charCount && positions.size < Display.Specs.lineCount) {
                    positions.add(0)
                }
            }
        }
        if (positions.lastIndex < Display.Specs.lineCount && positions.last() < Line.Specs.charCount) {
            currentCell = Cell.Cursor
        }
    }

    private fun deleteBackward() {
        if (positions.isEmpty() || positions.first() <= 0) {
            return
        }

        if (positions.last() < Line.Specs.charCount) {
            currentCell = Cell.Blank
        }
        positions[positions.lastIndex] -= 1
        if (positions.last() < 0) {
            positions.removeAt(positions.lastIndex)
            if (positions.last() >= Line.Specs.charCount) {
                positions[positions.lastIndex] = Line.Specs.charCount - 1
            }
        }
        if (positions.lastIndex < Display.Specs.lineCount && positions.last() < Line.Specs.charCount) {
            currentCell = Cell.Cursor
        }
    }
}

private fun Char.isAscii(): Boolean = (this.code < 127)
private fun Char.isNewLine(): Boolean = (this == '\n' || this == '\r')